File: typescript/src/giftcard/application/query/giftCardQueryService.ts
// Query service che espone i dati delle gift card all'API.

import { GiftCardId } from '../../domain/giftcard/giftCardId.js';
import type { GiftCardRepository } from '../../domain/ports/giftCardRepository.js';
import { type GiftCardDetails } from './giftCardDetails.js';

export class GiftCardQueryService {
  private readonly _repository: GiftCardRepository;

  constructor(repository: GiftCardRepository) {
    this._repository = repository;
  }

  findById(id: GiftCardId): GiftCardDetails | null {
    const card = this._repository.findById(id);
    if (card === null) {
      return null;
    }
    return {
      id: card.id().value.value,
      balance: card.balance(),
    };
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/query/GiftCardQueryService.java
package it.giannibombelli.wsc2026.giftcard.application.query;

import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;

import java.util.Optional;

public final class GiftCardQueryService {
    private final GiftCardRepository repository;

    public GiftCardQueryService(GiftCardRepository repository) {
        this.repository = repository;
    }

    public Optional<GiftCardDetails> findById(GiftCardId id) {
        return repository.findById(id)
            .map(this::toDetails);
    }

    private GiftCardDetails toDetails(GiftCard card) {
        return new GiftCardDetails(
            card.id().value(),
            card.balance()
        );
    }
}
